stages/org.osbuild.zipl-secex.inst

```diff
@@ -0,0 +1,234 @@
+        writer.add_file_from_memory(path, len(data), data)
+      archive.flush()
+      with open(archive.name, "rb") as f:
+        new_initrd.write(f.read())
+
+    new_initrd.flush()
+    return new_initrd
+
+
+def genprotimg(kernel, initrd, kopts, hostkey, sdboot):
+    with tempfile.NamedTemporaryFile(mode='w') as parmfile:
+      parmfile.write(kopts)
+      parmfile.flush()
+      subprocess.run(["/usr/bin/genprotimg",
+                      "-V",
+                      "--no-verify",
+                      "-i", kernel,
+                      "-r", initrd,
+                      "-p", parmfile.name,
+                      "-k", hostkey,
+                      "-o", sdboot],
+                     check=True)
+
+
+def open_dmverity(blocksize, partlabel, datapart, hashpart, mountpoint):
+    hash = subprocess.run(["/usr/sbin/veritysetup",
+                            "format", datapart, hashpart,
+                            "--data-block-size", f"{blocksize}"],
+                           check=True, capture_output=True)
+    for line in hash.stdout.decode().splitlines():
+      if line.startswith("Root hash:"):
+        hash = line.split(":", 1)[1].strip()
+        break
+
+    subprocess.run(["/usr/sbin/veritysetup",
+                    "open", datapart, partlabel, hashpart, hash],
+                   check=True)
+    subprocess.run(["mount", "-o", "ro", f"/dev/mapper/{partlabel}", mountpoint],
+                   check=True)
+    return hash
+
+
+def close_dmverity(partlabel, mountpoint):
+    subprocess.run(["umount", mountpoint], check=True)
+    subprocess.run(["/usr/sbin/veritysetup", "close", partlabel], check=True)
+
+
+def main(tree, paths, devices, options):
+    kver = options["kernel"]
+    append_kopts = options.get("kernel_opts_append", [])
+    blocksize = options.get("blocksize", 512)
+    build_id = options["build_id"]
+    location = options["location"]
+
+    hostkey = os.path.join(tree, options["hostkey"])
+    public_key = os.path.join(tree, options["ignition_gpg_pubkey"])
+
+    disk = os.path.join(paths["devices"], devices["disk"]["path"])
+    target = f"{paths["mounts"]}/se"
+
+    boot_fs = f"{disk}p3"
+    boothash_fs = f"{disk}p5"
+    root_fs = f"{disk}p4"
+    roothash_fs = f"{disk}p6"
+
+    with tempfile.TemporaryDirectory() as sysroot:
+      try:
+        roothash = open_dmverity(blocksize, "root", root_fs, roothash_fs, sysroot)
+        boothash = open_dmverity(blocksize, "boot", boot_fs, boothash_fs, f"{sysroot}/boot")
+
+        kernel, initrd, kopts = find_kernel(sysroot, kver)
+
+        secret_key = generate_gpg_keys(public_key, build_id)
+        sdboot = os.path.join(target, "sdboot")
+        with append_to_initrd(initrd, "/usr/lib/coreos/ignition.asc", secret_key) as initrd:
+          if append_kopts:
+            kopts = f"{kopts} {' '.join(append_kopts)}"
+          kopts = f"{kopts} rootfs.roothash={roothash}"
+          kopts = f"{kopts} bootfs.roothash={boothash}"
```

Comment:
would using `+=` here work better?

Reply:
IMO f-string literals are more easier to read